keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -232,27 +287,248 @@ demonstrate the interest in a KEP within the wider Kubernetes community.
 [experience reports]: https://github.com/golang/go/wiki/ExperienceReports
 -->
 
+Originally, Kubernetes and its scheduler only tracked CPU and RAM as
+resources for containers. Later, support for storage and discrete,
+countable per-node extended resources was added. The kubelet device plugin
+interface then made such local resources available to containers. But
+for many newer devices, this approach and the Kubernetes API for
+requesting these custom resources is too limited. This KEP may eventually
+address limitations of the current approach for the following use cases:
+
+- *Device initialization*: When starting a workload that uses
+  an accelerator like an FPGA, I’d like to have the accelerator
+  reconfigured or reprogrammed for the workload before the workload
+  itself starts. For security reasons, workloads should not be able to
```

Comment:
Any suggestion where this should go? How should this section be worded if security isn't mentioned?

Reply:
Security use-case could be something like this:
> Device initialization: Improves security by moving need for elevated device configuration privileges from workloads to kubelet driver plugin.  This reduces cluster attack surface and allows cluster admin better control of it.
>
> Limitation: with the current device plugin API, e.g. FPGA workloads need reconfigure / reprogram the device themselves, which can be done only as root user.

Another motivation could be performance:
> Device initialization: When device is available as a pre-programmed resource, workload startup is not delayed by the need to re-program it.

(3rd reason for doing it separately, could be enforcing of different device functionalities (available resources) at different times. For example, during day time, most of the devices could be programmed for real-time tasks, and during night (when real-time jobs are not expected) they could be changed to support batch jobs instead.  Or in MIG case, splitting the device different during day and night times.)

Reply:
I've focused on the security aspect.

I left out the part about "not delayed by the need to re-program it" because without support for immediate allocation this KEP doesn't really address this. For the same reason I also left out the 3rd point because it's not obvious how that matches the features of the KEP.

Reply:
> I've focused on the security aspect.

Err, you removed it...  Now the text does not mention security or elevated privileges at all, just "full HW access" which is very unspecific.

IMHO text should mention security impact explicitly, along lines of:
> Improves security by moving need for elevated device configuration privileges from workloads to kubelet driver plugin. This reduces cluster attack surface and allows cluster admin better control of it.

Reply:
> I’d like to have the accelerator reconfigured or reprogrammed without having to deploy my application
  with full hardware access.

I would also think that it's the cluster admin that cares more about apps not having (needing to have) elevated privileges (for obvious security reasons), rather than people deploying their apps (who like more freedom in what they do).

Reply:
I think it's a shared responsibility. What matters is that DRA needs to enable this.

I've added a comment saying that this is for security.

Reply:
Thanks, looks good now!   (please resolve)